Tristate checkbox frontend component (ZNTA-2116)

efloden opened the pull request:
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-2116

Created TriCheckbox component with indeterminate state.

TriCheckbox Storybook made available in storybook-frontend displaying:
- default component with truth table.
- usage in a checkbox group.
- custom styles applied through component props.

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.


The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?


----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zanata/zanata-platform/445)
<!-- Reviewable:end -->

davidmason commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/components.story.js, line 32 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn3jyMnJCotlOzyGTR:-Kpn3jyMnJCotlOzyGTS:b-p0viu8) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/components.story.js#L32)):*
> ```JavaScript
> require('./Tooltip/Tooltip.story.js')
> require('./Well/Well.story.js')
> require('./TriCheckbox/TriCheckbox.story')
> ```

Put in alphabetical order (I still didn't update the script to put these in the right place).

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.css, line 2 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn4Gg1aaL6_4VYrqb6:-Kpn4Gg1aaL6_4VYrqb7:bqj6gnx) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.css#L2)):*
> ```CSS
> 
> .tri-checkbox-story {
> ```

It looks like all these are all just for the story, and I can't see any styles for the `tri-checkbox` class that is used in the actual component.

I haven't really needed css classes for storybook components before so I haven't had a chance to consider how to approach it. I don't think we want storybook-only styles building into the production css (since they increase the transfer and load time of our css). If it were just trivial I would say put it as inline styles, but that won't work with the different states.

I think this file should be TriCheckbox.story.css since that is what it is for. Maybe you can just import it into the storybook directly with `import ./TriCheckbox.story.css` - if not, we can probably import it to the storybook css file in the storybook config.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 6 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6NmtCqh7Avvnxpf-:-Kpn6NmtCqh7Avvnxpf0:b-g0d6jg) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L6)):*
> ```JavaScript
> /**
>  * Checkbox with an intermediate state.
>  * TODO: Write 'tri-checkbox' style
> ```

Do you want to leave this `TODO` here, or implement that style?

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 21 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6tp_4JqtM69TLN9p:-Kpn6tpa_vJ__jJBQS-L:bxhftd3) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L21)):*
> ```JavaScript
>     const className = cx('tri-checkbox', this.props.className)
>     const { indeterminate, ...otherProps } = this.props
>     const hasIndeterminate = (nativeComponent) => {
> ```

This makes a new function instance every render cycle. Move it out of the render function onto the class.

I think you need 2 changes here - move this function to the class and have it store the ref (e.g. as `this.nativeComponent`) as well as set the value on it, and add `componentWillReceiveProps` where if `props.intermediate` is changing, you can set `this.nativeComponent.intermediate` with the new value (if it is present yet).

As it is, I think it will be creating a new function instance every render, and hammering the DOM with updates - generally not a good idea.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 17 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn8yZX6PJtpxHIIhx6:-Kpn8yZX6PJtpxHIIhx7:b-ma3y1o) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L17)):*
> ```JavaScript
>     checked: false,
>     indeterminate: false,
>     onClick: () => {}
> ```

I don't think there should be any `defaultProps` here. It should have the same contract as the real component, so only `intermediate` is required, and it should have no default value.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 34 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn9cGx9m9cx59IW7B7:-Kpn9cGx9m9cx59IW7B8:bneivep) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L34)):*
> ```JavaScript
>   makeIndeterminate = (event) => {
>     this.setState({checked: false, indeterminate: true})
>     this.props.onClick(event)
> ```

Does it really fire an `onClick` when you set the state to indeterminate? I wouldn't expect that behaviour, but if that is the behaviour it should be coming from the real component, not done by the wrapper (i.e. just let the real component call the `this.onClick` that you passed it.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 45 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnA4WDQQtE_ewKpUpl:-KpnA4WEE69R0v2_cfLk:bt1dyuo) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L45)):*
> ```JavaScript
>         />
>         <button onClick={this.makeIndeterminate}>Make indeterminate</button>
>         <h2>Truth Table</h2>
> ```

This is the wrong place for this. All this sort of display logic should go in a story. The wrapper is just for handling the state for demonstration purposes.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 107 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnAOqPYj8W-XgDhxfd:-KpnAOqPYj8W-XgDhxfe:b-imu88n) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L107)):*
> ```JavaScript
> }
> 
> class TriCheckboxGroup extends Component {
> ```

Move this to a separate file. It is ok to put the state wrapper component in the storybook, but generally not other components.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 212 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnAeZjL2K6HJcrlmbQ:-KpnAeZjL2K6HJcrlmbR:bds3gn4) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L212)):*
> ```JavaScript
> }
> 
> class FancyTriCheckbox extends Component {
> ```

Remove this class, just put the contents of the render method inside the story directly, like `.add('with custom styles', () => (<div>...`

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 11 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBRf0xaoTBZMeoTO6:-KpnBRf0xaoTBZMeoTO7:b-nh28mk) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L11)):*
> ```JavaScript
> 
> describe('TriCheckbox', () => {
>   // Markup tests should match truth table in storybook
> ```

Is this comment still relevant? Looks like it was a note from before tests were written.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 13 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBZWwdOQqtbe10Skm:-KpnBZWwdOQqtbe10Skn:b-3ykvrk) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L13)):*
> ```JavaScript
>   // Markup tests should match truth table in storybook
>   it('can render !checked && !indeterminate markup', () => {
>     const clickFun = function (e) {}
> ```

Let's use the `const callback = () => {}` approach here, like with the other new tests.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 74 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCP1HPJuHSPyzZkl7:-KpnCP1HPJuHSPyzZkl8:b8v53hh) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L74)):*
> ```JavaScript
>     }
>     expect(slothCheckbox.props.checked).toEqual(true)
>     expect(slothCheckbox.props.indeterminate).toEqual(false)
> ```

This is just looking at the props you passed in, which means the test can never fail no matter what the component is doing.

I think what you need to do here is make the click function store whatever it was called with, then check that value is what you expect or has been updated.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 76 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCB6aDid3posLUGC6:-KpnCB6aDid3posLUGC7:b-d2sipy) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L76)):*
> ```JavaScript
>     expect(slothCheckbox.props.indeterminate).toEqual(false)
>   })
> })
> ```

It doesn't look like any of these tests actually check the intermediate state. Can you query for the inner checkbox component and check that it has the right state? The test utils have tools you can use for that, I think we do it in some other tests too.

---

*[server/zanata-frontend/src/frontend/app/styles/style.less, line 4448 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnDKBse5RTu7M3kOGx:-KpnDKBse5RTu7M3kOGy:bx7ipzu) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/styles/style.less#L4448)):*
> ```Less
> 
> // Include CSS for tri-checkbox component
> @import "../components/TriCheckbox/index.css";
> ```

This would be fine, except that file is all storybook-only. Just adding a note here, because this will probably stay here if the styles for the component go in that file instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/components.story.js, line 32 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn3jyMnJCotlOzyGTR:-KpnLDc057eQz0gS-5p1:b-piup25) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/components.story.js#L32)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Put in alphabetical order (I still didn't update the script to put these in the right place).
</blockquote></details>

Moved this up.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.css, line 2 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn4Gg1aaL6_4VYrqb6:-KpnLPRl92ed9Iavx8fA:b-rfdz3) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.css#L2)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

It looks like all these are all just for the story, and I can't see any styles for the `tri-checkbox` class that is used in the actual component.

I haven't really needed css classes for storybook components before so I haven't had a chance to consider how to approach it. I don't think we want storybook-only styles building into the production css (since they increase the transfer and load time of our css). If it were just trivial I would say put it as inline styles, but that won't work with the different states.

I think this file should be TriCheckbox.story.css since that is what it is for. Maybe you can just import it into the storybook directly with `import ./TriCheckbox.story.css` - if not, we can probably import it to the storybook css file in the storybook config.
</blockquote></details>

Originally I made these styles apply to .tri-checkbox and they would be automatically applied to every component, but as they're just an example of passing styles for the storybook I gave them a separate name and left a TODO in the component to write the actual style.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 11 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBRf0xaoTBZMeoTO6:-KpnPRA9g3U8gfaMw2Ha:btz3z3z) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L11)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Is this comment still relevant? Looks like it was a note from before tests were written.
</blockquote></details>

Meant this as a reminder that if the indeterminate logic changes these tests will need to change as well.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 13 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBZWwdOQqtbe10Skm:-KpnOp7wywjUIPaF28S6:b-5bg3xt) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L13)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Let's use the `const callback = () => {}` approach here, like with the other new tests.
</blockquote></details>

Right, moved to a single declaration at the top of the file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

davidmason commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 11 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBRf0xaoTBZMeoTO6:-KpqyF7Czk1bHg7GvP3d:btj1rl9) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L11)):*
<details><summary><i>Previously, efloden (Earl Floden) wrote…</i></summary><blockquote>

Meant this as a reminder that if the indeterminate logic changes these tests will need to change as well.
</blockquote></details>

The tests themselves act as a big red reminder of that, so the comment is not needed. I think the YAGNI principle applies here, because we shouldn't expect to be changing the intermediate logic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 0 of 5 files reviewed at latest revision, 14 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 21 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6tp_4JqtM69TLN9p:-KprDcdiXv4Y-txnSMoi:b-ncam6q) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L21)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This makes a new function instance every render cycle. Move it out of the render function onto the class.

I think you need 2 changes here - move this function to the class and have it store the ref (e.g. as `this.nativeComponent`) as well as set the value on it, and add `componentWillReceiveProps` where if `props.intermediate` is changing, you can set `this.nativeComponent.intermediate` with the new value (if it is present yet).

As it is, I think it will be creating a new function instance every render, and hammering the DOM with updates - generally not a good idea.
</blockquote></details>

Moved outside of the render function and supported with componentWillRecieveProps

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 17 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn8yZX6PJtpxHIIhx6:-KprAjzfgV4493M7GyQC:byfdvnw) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L17)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

I don't think there should be any `defaultProps` here. It should have the same contract as the real component, so only `intermediate` is required, and it should have no default value.
</blockquote></details>

Removed default props from this wrapper - now a wrapper that just handles basic state of the component

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 34 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn9cGx9m9cx59IW7B7:-KprAyUCJmq5NSdQ2d4G:b-sgfe0t) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L34)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Does it really fire an `onClick` when you set the state to indeterminate? I wouldn't expect that behaviour, but if that is the behaviour it should be coming from the real component, not done by the wrapper (i.e. just let the real component call the `this.onClick` that you passed it.
</blockquote></details>

Removed it here, not needed for the wrapper

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 45 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnA4WDQQtE_ewKpUpl:-KprBexnVOX69VZFHloC:b-xvz0cz) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L45)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This is the wrong place for this. All this sort of display logic should go in a story. The wrapper is just for handling the state for demonstration purposes.
</blockquote></details>

Moved this into the story, also pulled out the button/checkbox example with its own WithButton class as the 'default' story

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 212 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnAeZjL2K6HJcrlmbQ:-KprC4Tp-280OMfgN7jj:b-4kg1yi) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L212)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Remove this class, just put the contents of the render method inside the story directly, like `.add('with custom styles', () => (<div>...`
</blockquote></details>

Moved this into the story

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 11 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnBRf0xaoTBZMeoTO6:-KprCAcvGEmUY_NFZE3p:beaa9n5) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L11)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

The tests themselves act as a big red reminder of that, so the comment is not needed. I think the YAGNI principle applies here, because we shouldn't expect to be changing the intermediate logic.
</blockquote></details>

Okay, removed this - IAGNI  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 0 of 6 files reviewed at latest revision, 14 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 107 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnAOqPYj8W-XgDhxfd:-KprEuPa1aSHw8UZxN-V:bsb7b1m) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L107)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Move this to a separate file. It is ok to put the state wrapper component in the storybook, but generally not other components.
</blockquote></details>

Moved to a separate file under TriCheckbox - could move it into its own TriCheckBoxGroup component?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

davidmason commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 21 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6tp_4JqtM69TLN9p:-KprW3StsFMk34tPN5_Y:b-4jqc4e) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L21)):*
<details><summary><i>Previously, efloden (Earl Floden) wrote…</i></summary><blockquote>

Moved outside of the render function and supported with componentWillRecieveProps
</blockquote></details>

Better change the name from `hasIntermediate` to something like `setNativeComponent`

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 34 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprWHRFMDTgWPYjMnGv:-KprWHRFMDTgWPYjMnGw:bmq37lm) ([raw file](https://github.com/zanata/zanata-platform/blob/c2445e3b7b877bc8445f7e58a1d0caa12c4a7d93/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L34)):*
> ```JavaScript
>   render () {
>     const className = cx('tri-checkbox', this.props.className)
>     const { ...otherProps } = this.props
> ```

This is making a copy of *all* the props, since `...` will only skip keys that appear before it and there are none before it. You need to put all the props you are using in front of the `...otherProps` to avoid errors about invalid props being passed to the input component. 

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 61 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprXrv6sZ-vWkESBCw1:-KprXrv6sZ-vWkESBCw2:bij3sh9) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L61)):*
> ```JavaScript
>                 checked={false}
>                 indeterminate={false}
>                 onChange={this.onClick}
> ```

You can still pass in `onChange` events here, just make them `action('onChange')` so they are shown in the storybook action log.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 107 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnAOqPYj8W-XgDhxfd:-KprY6_F_MTJf_XykrgP:bm37toz) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L107)):*
<details><summary><i>Previously, efloden (Earl Floden) wrote…</i></summary><blockquote>

Moved to a separate file under TriCheckbox - could move it into its own TriCheckBoxGroup component?
</blockquote></details>

I think it is ok under TriCheckBox for now, if it still needs more development and testing. It should just be moved to its own directory before we start using it in the app.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 32 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprX0pB8t7PF1HDqSi4:-KprX0pB8t7PF1HDqSi5:b-bpwa6m) ([raw file](https://github.com/zanata/zanata-platform/blob/c2445e3b7b877bc8445f7e58a1d0caa12c4a7d93/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L32)):*
> ```JavaScript
>           checked={this.state.checked}
>           indeterminate={this.state.indeterminate}
>           onChange={this.checkboxChanged}
> ```

You should still accept `onChange` in the wrapper, and call it inside `checkboxChanged`. The wrapper should just be spying on the action to see what state change it needs to make, but not take over it completely.

---

*[server/zanata-frontend/src/frontend/app/styles/style.less, line 4448 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnDKBse5RTu7M3kOGx:-KprYfZViXPFQ8HsHFd-:b1bbcds) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/styles/style.less#L4448)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This would be fine, except that file is all storybook-only. Just adding a note here, because this will probably stay here if the styles for the component go in that file instead.
</blockquote></details>

This should not be removed, since the styles are now applying to `tri-checkbox` so they will apply in the real app.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 3 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 21 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6tp_4JqtM69TLN9p:-Kprbk2nnr5Cisgcrh8p:b-y6uvo1) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L21)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Better change the name from `hasIntermediate` to something like `setNativeComponent`
</blockquote></details>

Changed this

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 34 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprWHRFMDTgWPYjMnGv:-KprbGizjzj5peRkf4dG:buho9ve) ([raw file](https://github.com/zanata/zanata-platform/blob/c2445e3b7b877bc8445f7e58a1d0caa12c4a7d93/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L34)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This is making a copy of *all* the props, since `...` will only skip keys that appear before it and there are none before it. You need to put all the props you are using in front of the `...otherProps` to avoid errors about invalid props being passed to the input component. 
</blockquote></details>

Thank you, I was trying to find the cause of that error, had to turn of the linter for unused vars to allow me to pull out  ```indeterminate``` here

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 61 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprXrv6sZ-vWkESBCw1:-Kprq0_88i-r2WYFrUdd:b-rp6z67) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L61)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

You can still pass in `onChange` events here, just make them `action('onChange')` so they are shown in the storybook action log.
</blockquote></details>

Passing it down as ```onChange={action('onChange')``` to the WithButton and TriCheckbox wrappers

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 32 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KprX0pB8t7PF1HDqSi4:-KprpkwieHIgjiEPoeOo:b-5gkjna) ([raw file](https://github.com/zanata/zanata-platform/blob/c2445e3b7b877bc8445f7e58a1d0caa12c4a7d93/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L32)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

You should still accept `onChange` in the wrapper, and call it inside `checkboxChanged`. The wrapper should just be spying on the action to see what state change it needs to make, but not take over it completely.
</blockquote></details>

Okay, I'm passing it down as ```onChange={action('onChange')``` to the WithButton and TriCheckbox wrappers

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 74 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCP1HPJuHSPyzZkl7:-KpsC8YvrEB_V4pwykSH:b1p5k7d) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L74)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This is just looking at the props you passed in, which means the test can never fail no matter what the component is doing.

I think what you need to do here is make the click function store whatever it was called with, then check that value is what you expect or has been updated.
</blockquote></details>

Testing the click method here as well now.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 76 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCB6aDid3posLUGC6:-KpsCJ14i5nSVkg4Ai5Y:b-7mln6c) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L76)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

It doesn't look like any of these tests actually check the intermediate state. Can you query for the inner checkbox component and check that it has the right state? The test utils have tools you can use for that, I think we do it in some other tests too.
</blockquote></details>

Testing the intermediate state for each of these  using TestUtils.renderIntoDocument now.

---

*[server/zanata-frontend/src/frontend/app/styles/style.less, line 4448 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnDKBse5RTu7M3kOGx:-Kprni4b0XZkISLVMfF8:barccpq) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/styles/style.less#L4448)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This should not be removed, since the styles are now applying to `tri-checkbox` so they will apply in the real app.
</blockquote></details>

Okay, I will re-work the styles to match the real app and leave this here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 1 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js, line 6 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-Kpn6NmtCqh7Avvnxpf-:-KqKu5X23SvwqhIClTI_:b-jso7m2) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/index.js#L6)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Do you want to leave this `TODO` here, or implement that style?
</blockquote></details>

Implemented a blue checkbox style to match the Zanata theme using @color-light from the variables.less file

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

davidmason commented:
Reviewed 4 of 7 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 27 at r4](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqLCohhFgqvn-8JVZ_X:-KqLCohhFgqvn-8JVZ_Y:bk7ej23) ([raw file](https://github.com/zanata/zanata-platform/blob/41ff2fd4245b140d341d01a70c95e73f85e1a43b/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L27)):*
> ```JavaScript
>   checkboxChanged = (event) => {
>     this.setState({checked: event.target.checked, indeterminate: false})
>     this.props.onChange()
> ```

`this.props.onChange(event)` so it shows the right thing in action logging

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 74 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCP1HPJuHSPyzZkl7:-KqLGcf7Y_dHvR2I_IlS:b-blwg48) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L74)):*
<details><summary><i>Previously, efloden (Earl Floden) wrote…</i></summary><blockquote>

Testing the click method here as well now.
</blockquote></details>

These 2 lines are still not testing anything, better to remove them since they give the false impression that something is being checked.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxGroup.js, line 88 at r4](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqLGsvNSGMy2SPGsI_H:-KqLGsvNSGMy2SPGsI_I:b-fm6p5) ([raw file](https://github.com/zanata/zanata-platform/blob/41ff2fd4245b140d341d01a70c95e73f85e1a43b/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxGroup.js#L88)):*
> ```JavaScript
>             indeterminate={false}
>             onChange={this.subCheckboxClick}
>             useDefaultStyle={false}
> ```

The group component should be able to use the default checkbox styles. Not a blocker for this review though since it is just an extra.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js, line 27 at r4](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqLCohhFgqvn-8JVZ_X:-KqLQkUC73ijkvNl6nCJ:baqstip) ([raw file](https://github.com/zanata/zanata-platform/blob/41ff2fd4245b140d341d01a70c95e73f85e1a43b/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.story.js#L27)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

`this.props.onChange(event)` so it shows the right thing in action logging
</blockquote></details>

Passing the event to these now.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js, line 74 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KpnCP1HPJuHSPyzZkl7:-KqLQouu9ewbekFHCPFa:b-iu3hlp) ([raw file](https://github.com/zanata/zanata-platform/blob/e48450e25cb5b84fd8e8867b1c66184e11110917/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckbox.test.js#L74)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

These 2 lines are still not testing anything, better to remove them since they give the false impression that something is being checked.
</blockquote></details>

Removed these two tests.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxGroup.js, line 88 at r4](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqLGsvNSGMy2SPGsI_H:-KqLQuQTYYyHuHpS7aHQ:bqulm7k) ([raw file](https://github.com/zanata/zanata-platform/blob/41ff2fd4245b140d341d01a70c95e73f85e1a43b/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxGroup.js#L88)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

The group component should be able to use the default checkbox styles. Not a blocker for this review though since it is just an extra.
</blockquote></details>

They can use the default style, enabled it on them so it shows in the group storybook as well now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

davidmason commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js, line 1 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqMAozvvFpFDV6e4qw8:-KqMAozvvFpFDV6e4qw9:b-wpzcds) ([raw file](https://github.com/zanata/zanata-platform/blob/642a6ca8267d18e846507760e30a561fdf1872cf/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js#L1)):*
> ```JavaScript
> import React, { Component } from 'react'
> ```

This would be better in a separate PR since your branch is already reviewed.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js, line 11 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqM8KJsTQ9C4AZI7tBp:-KqM8KJsTQ9C4AZI7tBq:b-nqf134) ([raw file](https://github.com/zanata/zanata-platform/blob/642a6ca8267d18e846507760e30a561fdf1872cf/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js#L11)):*
> ```JavaScript
>     title: PropTypes.string.isRequired,
>     checked: PropTypes.bool.isRequired,
>     indeterminate: PropTypes.bool.isRequired,
> ```

Do not pass in intermediate and checked, since the options should determine these values.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js, line 16 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqM8UY5Cq5UXhxKQaXK:-KqM8UY5Cq5UXhxKQaXL:b-dkjx52) ([raw file](https://github.com/zanata/zanata-platform/blob/642a6ca8267d18e846507760e30a561fdf1872cf/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js#L16)):*
> ```JavaScript
>       checked: PropTypes.bool,
>       name: PropTypes.string
>     })),
> ```

I suggest you follow the React input component convention, where they have `value` and `defaultValue` that determine whether this should be a controlled component. In this case have `options` and `defaultOptions`:

- if `options` and `defaultOptions` are both present, it is an error - fail fast and throw an exception. Developers should not pass both in.
- if `defaultOptions` is provided, set `this.controlled = false`, copy it to state and manage it all in `this.state.options`
- if `options` is provided, set `this.controlled = true`, don't handle any state yourself, just render the current state based on props. You still fire the callbacks, but don't do anything with state.
- if someone switches between `options` and `defaultOptions`, it is probably a mistake so print a warning.

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js, line 40 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqM7ErfyJwvo0GafWd3:-KqM7ErfyJwvo0GafWd4:b-qp81op) ([raw file](https://github.com/zanata/zanata-platform/blob/642a6ca8267d18e846507760e30a561fdf1872cf/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js#L40)):*
> ```JavaScript
>     this.setState((prevState, props) => ({
>       checked: checked,
>       indeterminate: false,
> ```

This should not store `checked` or `intermediate` in its state. You can always infer those from `this.props.options` (or `this.state.options`) so it is duplicate data (and a potential source of bugs). Just render checked based on something like `const checked = every(options, ({ checked } => checked)` and `const intermediate = !checked && some(options, ({ checked }) => checked)` (see lodash `_.every` and `_.some`)

---

*[server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js, line 68 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-KqM9y0359wGuzyhglws:-KqM9y0359wGuzyhglwt:b-x3u5ui) ([raw file](https://github.com/zanata/zanata-platform/blob/642a6ca8267d18e846507760e30a561fdf1872cf/server/zanata-frontend/src/frontend/app/components/TriCheckbox/TriCheckboxDropdown.js#L68)):*
> ```JavaScript
>       }) !== undefined)
>     }))
>     this.props.onClick(event)
> ```

This will be tricky to use - someone has to dig into the event.target and try to figure out which option was clicked. I suggest calling a callback with something like `callback(name, event.target.checked)` so that there is some way for the application to tell what was clicked.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445)*
<!-- Sent from Reviewable.io -->

davidmason commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -(Under review) +Reviewed 

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/445#-:-KqQO4tnBWAcqwbMDWW0:bns32bj)*
<!-- Sent from Reviewable.io -->

spathare21 commented:
@efloden I did not see the any changed with checkbox througout the application after mergin this PR

efloden commented:
@spathare21 the checkboxes in the application should not change with this PR, but it is planned to replace them with it after it has been merged - this component can only be observed in the storybook-frontend in this PR

spathare21 commented:
ok thanks @efloden then marking it as verified as build pass and everything looks good
